### PR TITLE
{Core} Fix `random_config_dir` when `AZURE_CONFIG_DIR` is set

### LINKED ISSUE #28668: When `AZURE_CONFIG_DIR` is set, login context gets cleared by setting `random_config_dir=True`
ADO's Azure CLI task contains this step:

```
Setting AZURE_CONFIG_DIR env variable to: /mnt/vss/_work/_temp/.azclitask
```

In Knack, the config dir specified by env var takes higher precedence:

https://github.com/microsoft/knack/blob/e0c14114aea5e4416c70a77623e403773aba73a8/knack/config.py#L44

```py
        self.config_dir = os.environ.get('{}CONFIG_DIR'.format(env_var_prefix), default_config_dir)
```

So when `AZURE_CONFIG_DIR` is set, setting `config_dir` becomes a no-op for `AzCli`:

https://github.com/Azure/azure-cli/blob/c087891017008dd22c49d14742dc7d6480212cfd/src/azure-cli-core/azure/cli/core/mock.py#L44-L46

`self.config.config_dir` in this case is the value of `AZURE_CONFIG_DIR` (`/mnt/vss/_work/_temp/.azclitask`).

During `tearDown`, directory specified by `AZURE_CONFIG_DIR` gets deleted:

https://github.com/Azure/azure-cli/blob/b223dd9d4379d5c633a9f4dba4941ae791531126/src/azure-cli-testsdk/azure/cli/testsdk/base.py#L139-L144

This causes failure in the following `azdev test --live` command:

```
self = <azure.cli.core._profile.Profile object at 0x7f3d731794b0>
subscription = None

    def get_subscription(self, subscription=None):  # take id or name
        subscriptions = self.load_cached_subscriptions()
        if not subscriptions:
>           raise CLIError(_AZ_LOGIN_MESSAGE)
E           knack.util.CLIError: Please run 'az login' to setup account.

src/azure-cli-core/azure/cli/core/_profile.py:546: CLIError
```

Introduced by #25689

### PR DESCRIPTION
**Pull request by bebound**
Resolve #28668

Knack ignores the `config_dir` param when `AZURE_CONFIG_DIR` env is set, and other functions may call `get_config_dir` directly.
So this PR patches `AZURE_CONFIG_DIR` env when `random_config_dir=true` during test

### CODE REVIEW

**Comment by yonzhan** [comment]
Core

File: src/azure-cli-core/azure/cli/core/mock.py [hunk 1]
```diff
@@ -26,6 +27,10 @@ def __init__(self, commands_loader_cls=None, random_config_dir=False, **kwargs):
 
         if random_config_dir:
             config_dir = os.path.join(GLOBAL_CONFIG_DIR, 'dummy_cli_config_dir', random_string())
+            # Knack prioritizes the AZURE_CONFIG_DIR env over the config_dir param, and other functions may call
+            # get_config_dir directly. We need to set the env to make sure the config_dir is used.
+            self.env_patch = patch.dict('os.environ', {'AZURE_CONFIG_DIR': config_dir})
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/mock.py:32]: `os.environ` should suffice:

```suggestion
            self.env_patch = patch.dict(os.environ, {'AZURE_CONFIG_DIR': config_dir})
```

File: src/azure-cli-core/azure/cli/core/mock.py [hunk 2]
```diff
@@ -38,12 +43,10 @@ def __init__(self, commands_loader_cls=None, random_config_dir=False, **kwargs):
                             shutil.copy(os.path.join(GLOBAL_CONFIG_DIR, file), config_dir)
                         except FileNotFoundError:
                             pass
-        else:
-            config_dir = GLOBAL_CONFIG_DIR
 
         super(DummyCli, self).__init__(
             cli_name='az',
-            config_dir=config_dir,
+            config_dir=GLOBAL_CONFIG_DIR,
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/mock.py:49]: This is what it looks like before #25689.

**Review by jiasli — APPROVED** [review]

File: src/azure-cli-core/azure/cli/core/mock.py [hunk 3]
```diff
@@ -26,6 +27,10 @@ def __init__(self, commands_loader_cls=None, random_config_dir=False, **kwargs):
 
         if random_config_dir:
             config_dir = os.path.join(GLOBAL_CONFIG_DIR, 'dummy_cli_config_dir', random_string())
+            # Knack prioritizes the AZURE_CONFIG_DIR env over the config_dir param, and other functions may call
+            # get_config_dir directly. We need to set the env to make sure the config_dir is used.
```
Comment (bebound) [inline src/azure-cli-core/azure/cli/core/mock.py:31]: > self.config_dir = os.environ.get('{}CONFIG_DIR'.format(env_var_prefix), default_config_dir)
-- https://github.com/microsoft/knack/blob/e0c14114aea5e4416c70a77623e403773aba73a8/knack/config.py#L44

**Pull request merged**